components/brave_wallet/browser/zcash/zcash_wallet_service.cc

```diff
@@ -321,11 +342,109 @@ void ZCashWalletService::GetUtxos(const std::string& chain_id,
+    if (!address_map.contains(input.utxo_address)) {
+      return false;
+    }
+    auto& key_id = address_map.at(input.utxo_address);
+
+    auto pubkey = keyring_service_->GetZCashPubKey(account_id, key_id);
+    if (!pubkey) {
+      return false;
+    }
+
+    auto signature_digest =
+        ZCashSerializer::CalculateSignatureDigest(tx, input);
+
+    auto signature = keyring_service_->SignMessageByZCashKeyring(
+        account_id, key_id,
+        base::make_span<32>(signature_digest.begin(), signature_digest.end()));
+    if (!signature) {
+      return false;
+    }
+
+    BtcLikeSerializerStream stream(&input.script_sig);
+    stream.PushVarInt(signature.value().size() + 1);
+    stream.PushBytes(signature.value());
+    stream.Push8AsLE(tx.sighash_type());
+    stream.PushSizeAndBytes(pubkey.value());
+  }
+
+  return true;
+}
+
 void ZCashWalletService::SignAndPostTransaction(
     const std::string& chain_id,
     const mojom::AccountIdPtr& account_id,
     ZCashTransaction zcash_transaction,
-    SignAndPostTransactionCallback callback) {}
+    SignAndPostTransactionCallback callback) {
+  zcash_rpc_->GetLatestBlock(
+      chain_id,
+      base::BindOnce(
+          &ZCashWalletService::OnResolveLastBlockHeightForSendTransaction,
+          weak_ptr_factory_.GetWeakPtr(), chain_id, account_id.Clone(),
+          std::move(zcash_transaction), std::move(callback)));
+}
+
+void ZCashWalletService::OnResolveLastBlockHeightForSendTransaction(
+    const std::string& chain_id,
+    const mojom::AccountIdPtr& account_id,
+    ZCashTransaction zcash_transaction,
+    SignAndPostTransactionCallback callback,
+    base::expected<zcash::BlockID, std::string> result) {
+  if (!result.has_value()) {
+    std::move(callback).Run("", std::move(zcash_transaction),
+                            "Couldn't obtain last block");
+    return;
+  }
+
+  zcash_transaction.set_expiry_height(result->height() +
+                                      kDefaultBlockHeightDelta);
+
+  if (!SignTransactionInternal(zcash_transaction, account_id)) {
+    std::move(callback).Run("", std::move(zcash_transaction),
+                            "Couldn't sign transaciton");
+    return;
+  }
+
+  auto tx = ZCashSerializer::SerializeRawTransaction(zcash_transaction);
+  std::string as_string(reinterpret_cast<char*>(tx.data()), tx.size());
+  zcash_rpc_->SendTransaction(
+      chain_id, as_string,
+      base::BindOnce(&ZCashWalletService::OnSendTransactionResult,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback),
+                     std::move(zcash_transaction)));
+}
+
+void ZCashWalletService::OnSendTransactionResult(
+    SignAndPostTransactionCallback callback,
+    ZCashTransaction tx,
+    base::expected<zcash::SendResponse, std::string> result) {
+  if (result.has_value() && result->errorcode() == 0) {
+    auto tx_id = ZCashSerializer::CalculateTxIdDigest(tx);
+    auto tx_id_hex = ToHex(tx_id);
+    CHECK(tx_id_hex.starts_with("0x"));
```

Comment:
just confirming the only thing this does is return the TxId in hex format? If so, no concerns about rule of two here even though it's relying on the same `SendTransaction` call because the grammar here is simple.

Reply:
Do you mean RPC response for SendTransaction call? It returns just confirmation status and we calculate TxId on our side.